packages/arb-token-bridge-ui/src/components/TransferPanel/TransferPanelMainInput.tsx

```diff
@@ -140,34 +146,85 @@ export type TransferPanelMainInputProps =
     maxButtonOnClick: React.ButtonHTMLAttributes<HTMLButtonElement>['onClick']
     value: string
     tokenButtonOptions?: TokenButtonOptions
+    maxAmount: string | undefined
+    isMaxAmount: boolean
   }
 
-export function TransferPanelMainInput(props: TransferPanelMainInputProps) {
-  const { errorMessage, maxButtonOnClick, tokenButtonOptions, ...rest } = props
+export const TransferPanelMainInput = React.memo(
+  (props: TransferPanelMainInputProps) => {
+    const {
+      errorMessage,
+      maxButtonOnClick,
+      tokenButtonOptions,
+      onChange,
+      maxAmount,
+      value,
+      isMaxAmount,
+      ...rest
+    } = props
+    const [localValue, setLocalValue] = useState(value)
+
+    useEffect(() => {
+      if (!isMaxAmount || !maxAmount) {
+        return
+      }
 
-  return (
-    <>
-      <div
-        className={twMerge(
-          'flex flex-row rounded border bg-black/40 shadow-2',
-          errorMessage
-            ? 'border-brick text-brick'
-            : 'border-white/30 text-white'
-        )}
-      >
-        <TokenButton options={tokenButtonOptions} />
+      /**
+       * On first render, maxAmount is not defined, once we receive max amount value, we set the localValue
+       * If user types anything before we receive the amount, isMaxAmount is set to false in the parent
+       */
+      setLocalValue(maxAmount)
+    }, [isMaxAmount, maxAmount])
+
+    const handleMaxButtonClick: React.MouseEventHandler<HTMLButtonElement> =
+      useCallback(
+        e => {
+          maxButtonOnClick?.(e)
+          if (maxAmount) {
+            setLocalValue(maxAmount)
+          }
+        },
+        [maxAmount, maxButtonOnClick]
+      )
+
+    const handleInputChange: ChangeEventHandler<HTMLInputElement> = useCallback(
+      e => {
+        setLocalValue(sanitizeAmountQueryParam(e.target.value))
```

Comment:
use `truncateExtraDecimals` to truncate the extra length

see https://github.com/OffchainLabs/arbitrum-token-bridge/blob/master/packages/arb-token-bridge-ui/src/hooks/TransferPanel/useSetInputAmount.ts